paparazzi/paparazzi/src/main/java/app/cash/paparazzi/accessibility/ComposeA11yExtension.kt

```diff
@@ -0,0 +1,127 @@
+
+/**
+ * Paparazzi Render Extension that collects Accessibility information for Compose
+ * hierarchies.
+ *
+ * Currently captures Content Description, State Description, On Click, Role, Disabled,
+ * Heading, Custom Actions, Text, and Progress.  These are saved as AccessibilityState.Element in
+ * the [accessibilityState] list.
+ */
+class ComposeA11yExtension : RenderExtension {
+  lateinit var accessibilityState: AccessibilityState
+
+  private lateinit var rootForTest: RootForTest
+
+  init {
+    ViewRootForTest.onViewCreatedCallback = { viewRoot ->
+      viewRoot.view.addOnAttachStateChangeListener(object : View.OnAttachStateChangeListener {
+        override fun onViewAttachedToWindow(p0: View) {
+          // Grab the AndroidComposeView using the same hook that createComposeRule uses
+          // Note: It isn't usefully populated at this point.
+          rootForTest = p0 as RootForTest
+        }
+
+        override fun onViewDetachedFromWindow(p0: View) {
+        }
+      })
+    }
+  }
+
+  private fun processAccessibleChildren(
+    p0: SemanticsNode,
+    fn: (AccessibilityState.Element) -> Unit
+  ) {
+    val contentDescription = p0.config.getOrNull(SemanticsProperties.ContentDescription)
+    val stateDescription = p0.config.getOrNull(SemanticsProperties.StateDescription)
+    val onClickLabel = p0.config.getOrNull(SemanticsActions.OnClick)?.label
+    val role = p0.config.getOrNull(SemanticsProperties.Role)?.toString()
+    val disabled = p0.config.getOrNull(SemanticsProperties.Disabled) != null
+    val heading = p0.config.getOrNull(SemanticsProperties.Heading) != null
+    val customActions = p0.config.getOrNull(SemanticsActions.CustomActions)
+    val text = p0.config.getOrNull(SemanticsProperties.Text)
+    val progress = p0.config.getOrNull(SemanticsProperties.ProgressBarRangeInfo)
+    val hasProgressAction = p0.config.getOrNull(SemanticsActions.SetProgress) != null
+
+    if (contentDescription != null || stateDescription != null || onClickLabel != null || role != null || progress != null || text != null) {
+      val position = p0.boundsInRoot.toAndroidRect()
+      val touchBounds = p0.touchBoundsInRoot.toAndroidRect()
+      fn(
+        AccessibilityState.Element(
+          position,
+          if (touchBounds != position) touchBounds else null,
+          text?.map { it.toString() },
+          contentDescription,
+          stateDescription,
+          onClickLabel,
+          role,
+          disabled,
+          heading,
+          customActions?.map { AccessibilityState.CustomAction(label = it.label) },
+          progress?.let {
+            AccessibilityState.Progress(it.current, it.range, it.steps, hasProgressAction)
+          }
+        )
+      )
+    }
+
+    p0.children.forEach {
+      processAccessibleChildren(it, fn)
+    }
+  }
+
+  override fun renderView(contentView: View): View {
+    val composeView = (contentView as ViewGroup).getChildAt(0) as ComposeView
```

Comment:
this implementation expects the root view to be a `ComposeView`. It doesn't fix issues where you have a view hierarchy with mixed legacy views and compose views.

Reply:
I'll test with this scenario.  I'm not sure we can easily support this, I might end up making it clear what is supported.

The A11y model for Compose seems different from Views, so unless I changed this to work as an accessibility service, it might not be possible to mix (I'll confirm in a test). I think that still captures the final state of most apps.  Either completely views, or completely compose.

Reply:
> I think that still captures the final state of most apps. Either completely views, or completely compose.

Doesn't this exclude apps that are incrementally migrating to compose ui?

Reply:
I unfortunately don't have a good answer

Reply:
I unfortunately don't have a good answer, sorry!

Reply:
Ya, I was able to handle this case in this pr. https://github.com/cashapp/paparazzi/pull/702 
It takes a lot from the work in this PR.

Reply:
A big thank you for taking this over.